### PR TITLE
Refactor FirestoreCtrl into Modular Functions for Improved Testability and Maintainability

### PR DESCRIPTION
**Pull request by EduardoNeville**
This PR refactors the `FirestoreCtrl` class by splitting it into multiple modules: `GetFirestoreCtrl`, `SetFirestoreCtrl`, `TypeFirestoreCtrl`, and `LocalStorageCtrl`. The goal is to improve the modularity, readability, and testability of the codebase by separating concerns between data retrieval, data manipulation, type definitions, and local storage operations. 

Additionally, all relevant tests have been updated to reflect these changes, incorporating best practices for mocking nested functions and promises as per the following references:

- [Mock a nested function](https://stackoverflow.com/questions/73913544/how-to-mock-inner-function-while-testing-the-outer-function-in-jest)
- [Mock a Promise](https://stackoverflow.com/questions/56285530/how-to-create-jest-mock-function-with-promise)
- [Mock a returned void](https://stackoverflow.com/questions/68183112/in-jest-how-do-i-mock-a-promise-of-void)

#### a. **`GetFirestoreCtrl.tsx`**

- Contains functions responsible for fetching data from Firestore.
- Examples include `getUser`, `getChallenge`, `getCommentsOf`, etc.

#### b. **`SetFirestoreCtrl.tsx`**

- Contains functions responsible for writing or updating data in Firestore.
- Examples include `createUser`, `setName`, `updateLikesOf`, etc.

#### c. **`TypeFirestoreCtrl.tsx`**

- Houses all the type definitions (`DBUser`, `DBChallenge`, `DBComment`, etc.).
- Centralizes type definitions for consistency across the codebase.

#### d. **`LocalStorageCtrl.tsx`**

- Manages local storage operations, especially when offline.
- Handles caching data locally and uploading it when connectivity is restored.
- Functions include `storeImageLocally`, `uploadStoredImages`, etc.

### 2. Updating Application Code

- **Removed** all instances where `FirestoreCtrl` was imported and instantiated.
- **Replaced** method calls like `firestoreCtrl.getUser(uid)` with direct function calls `getUser(uid)`.
- **Updated Imports**: Imported functions directly from their respective modules.
  ```diff
  - import FirestoreCtrl from "@/src/models/firebase/FirestoreCtrl";
  + import { getUser } from "@/src/models/firebase/GetFirestoreCtrl";
  ```
- **Adjusted Function Signatures**: Removed `firestoreCtrl` parameters from functions where it was previously passed.

### 3. Modifying ViewModel and Component Files

- **ViewModels**: Updated to use the new modular functions.
  - **Example**: In `SetUsernameViewModel`, replaced `firestoreCtrl.setName` with `setName`.
  ```diff
  - await firestoreCtrl.setName(user.uid, username, setUser);
  + await setName(user.uid, username, setUser);
  ```
- **Components**: Adjusted to remove `firestoreCtrl` props and use the new functions.
  - **Example**: In `CameraContainer.tsx`, removed `firestoreCtrl` from props and updated function calls.

#### a. **Mocking Nested Functions and Promises**

- Employed techniques from the provided references to mock nested functions and promises effectively.
- **Example**: Mocking a promise-returning function:
  ```javascript
  jest.mock("@/src/models/firebase/GetFirestoreCtrl", () => ({
    getUser: jest.fn().mockResolvedValue(/* mock value */),
  }));
  ```

#### b. **Adjusting Jest Mocks**

- Replaced mocks of `FirestoreCtrl` with mocks of the specific functions from the new modules.
- Ensured that each test only mocks what it needs, aligning with unit testing best practices.

#### c. **Refactoring Test Suites**

- Updated all test files under `__tests__` to reflect the changes.
- **Removed** imports and references to `FirestoreCtrl`.
- **Updated** function calls and mock implementations accordingly.

### 5. Refactoring Types and Interfaces

- **Type Imports**: Changed imports of types like `DBUser` to come from `TypeFirestoreCtrl`.
  ```diff
  - import FirestoreCtrl, { DBUser } from "@/src/models/firebase/FirestoreCtrl";
  + import { DBUser } from "@/src/models/firebase/TypeFirestoreCtrl";
  ```
- Ensured all type definitions are consistent and centralized.

### 6. Updating Utility and Helper Functions

- In files like `Auth.tsx`, replaced calls to `firestoreCtrl` methods with direct function calls.
- **Example**: In `Auth.tsx`, updated `logInWithEmail` function:
  ```diff
  - await logInWithEmail(email, password, firestoreCtrl, navigation, setUser);
  + await logInWithEmail(email, password, navigation, setUser);
  ```

### 7. Cleaning Up Navigation and App Initialization

- Removed passing `firestoreCtrl` as a prop in navigation components.
- Adjusted initializations in `App.tsx` and `Navigation.tsx` to align with new structure.

## Testing

- **Unit Tests**: All unit tests have been updated and are passing.
- **Integration Tests**: Tested critical user flows to ensure no regressions.
- **Manual Testing**: Performed manual testing of features like user authentication, challenge creation, and friend management.

## References Used

- [Mock a nested function](https://stackoverflow.com/questions/73913544/how-to-mock-inner-function-while-testing-the-outer-function-in-jest)
- [Mock a Promise](https://stackoverflow.com/questions/56285530/how-to-create-jest-mock-function-with-promise)
- [Mock a returned void](https://stackoverflow.com/questions/68183112/in-jest-how-do-i-mock-a-promise-of-void)

## Rationale Behind Changes

- **Improved Modularity**: Separating `FirestoreCtrl` into logical modules makes the codebase easier to navigate and maintain.
- **Enhanced Testability**: Testing individual functions is more straightforward, and mocking is more precise.
- **Single Responsibility Principle**: Each module now has a clear responsibility, adhering to best coding practices.
- **Readability**: Developers can now understand and locate functionality faster, improving development speed and onboarding.

## Notes

- **Extensibility**: The new structure allows for easier addition of new features or services (e.g., adding new data retrieval functions in `GetFirestoreCtrl`).
- **Consistency**: Aligns with modern JavaScript and TypeScript practices of using modules and pure functions over classes where appropriate.

### CODE REVIEW

File: app/src/viewmodels/camera/CameraViewModel.tsx [hunk 1]
```diff
@@ -12,22 +12,22 @@ import {
...
- * @param firestoreCtrl : FirestoreCtrl object
  * @param navigation : navigation object
  * @returns : functions for the camera screen
  */
-export default function useCameraViewModel(
-  firestoreCtrl: FirestoreCtrl,
-  navigation: any,
-  route: any,
-) {
+export default function useCameraViewModel(navigation: any, route: any) {
```
Comment (Tristan-C789) [inline app/src/viewmodels/camera/CameraViewModel.tsx:30]: The types should be specified here.

File: app/src/viewmodels/camera/CameraViewModel.tsx [hunk 2]
```diff
@@ -137,17 +135,18 @@ export default function useCameraViewModel(
...
           alert("You need to be in the group's area to create a challenge");
           navigation.navigate("GroupScreen", { currentGroup: group });
           return;
         }
       }
 
-      const imageId = await firestoreCtrl.uploadImage(picture?.uri);
+      console.log("Picture URI: ", picture?.uri);
+      const imageId = await uploadImage(picture?.uri);
+      console.log("image id making challenge: ", imageId);
```
Comment (Tristan-C789) [inline app/src/viewmodels/camera/CameraViewModel.tsx:148]: You should probably remove those debug logs!

File: app/src/viewmodels/components/friends/ListOfFilteredUsersViewModel.tsx
Comment (Tristan-C789) [inline app/src/viewmodels/components/friends/ListOfFilteredUsersViewModel.tsx:1]: The variables with the same name (i.e. `isFriend` and `isFriendB`) shoud be renamed to something more explicit.

File: app/src/viewmodels/components/friends/RequestListViewModel.tsx [hunk 3]
```diff
@@ -1,22 +1,20 @@
+import {
+  acceptFriend,
+  rejectFriend,
+} from "@/src/models/firebase/SetFirestoreCtrl";
+
 /**
  * Request List with all users that have sent a friend request to the current user
```
Comment (Tristan-C789) [inline app/src/viewmodels/components/friends/RequestListViewModel.tsx:7]: This comment does not match the function I believe.

File: app/src/viewmodels/home/ProfileScreenViewModel.tsx [hunk 4]
```diff
@@ -1,25 +1,28 @@
 import { useEffect, useState } from "react";
 import { launchImageLibraryAsync } from "expo-image-picker";
-import FirestoreCtrl, { DBUser } from "../../models/firebase/FirestoreCtrl";
+import { DBUser } from "../../models/firebase/TypeFirestoreCtrl";
```
Comment (Tristan-C789) [inline app/src/viewmodels/home/ProfileScreenViewModel.tsx:3]: You should use the @ notation.

File: app/src/viewmodels/home/ProfileScreenViewModel.tsx [hunk 5]
```diff
@@ -30,19 +33,20 @@ export function useProfileScreenViewModel(
 } {
   const userIsGuest = user.name === "Guest";
 
-  const [name, setName] = useState<string>(user.name);
+  const [name, sName] = useState<string>(user.name);
```
Comment (Tristan-C789) [inline app/src/viewmodels/home/ProfileScreenViewModel.tsx:36]: Why the use of `sName` instead of `setName`?

Comment (Tristan-C789) [reply app/src/viewmodels/home/ProfileScreenViewModel.tsx:36]: Okay I realised why. I understand the reasoning, however one the method should be renamed.

File: app/src/views/components/friends/list_of_filtered_users.tsx [hunk 6]
```diff
@@ -29,7 +28,7 @@ export default function ListOfFilteredUsers({
           data={filteredUsers}
           keyExtractor={(item) => item.uid || Math.random().toString()}
           renderItem={({ item }) => {
-            const { isFriend, isRequested } = userStatuses[item.uid] || {
+            const { isFriendB, isRequestedB } = userStatuses[item.uid] || {
```
Comment (Tristan-C789) [inline app/src/views/components/friends/list_of_filtered_users.tsx:31]: Same here, this should be renamed while still making sure to not conflict with the Firestore one.

**Review by Tristan-C789 — CHANGES_REQUESTED** [review]
Overall this was an awesome refactor! 

It is well executed, with everything adapted. The tests are particularly well done! I requested a few changes here and there, but some do not really come from your PR (they should be easy fixes) or are superficial.

Thanks once again!

File: app/src/models/firebase/GetFirestoreCtrl.tsx [hunk 7]
```diff
@@ -0,0 +1,561 @@
...
+    throw new Error("Error getting user");
+  }
+}
+
+/**
+ * Get the url of an image
+ * @param id_picture The id of the image
+ * @returns The download URL of the image.
+ */
+export async function getImageUrl(id_picture: string): Promise<string> {
```
[the remark this answers is not in the document]
Comment (EduardoNeville) [reply app/src/models/firebase/GetFirestoreCtrl.tsx:61]: I will need guidance for the Memories as the use them. But will be deleted in the next pr

File: app/src/models/firebase/TypeFirestoreCtrl.tsx
[the remark this answers is not in the document]
Comment (EduardoNeville) [reply app/src/models/firebase/TypeFirestoreCtrl.tsx:1]: agreed, perhaps on another PR

File: app/src/viewmodels/camera/CameraViewModel.tsx (hunk 2 above)
Comment (EduardoNeville) [reply app/src/viewmodels/camera/CameraViewModel.tsx:148]: needed for the Memories screen

**Review by Tristan-C789 — APPROVED** [review]
Thanks for addressing the issues, LGTM !

**Pull request merged**